drivers/serial/uart_hvc_xen.c

```diff
@@ -0,0 +1,166 @@
+
+#include <device.h>
+#include <init.h>
+#include <kernel.h>
+#include <kernel_arch_interface.h>
+#include <sys/printk.h>
+
+static struct hvc_xen_data hvc_data = {0};
+
+static int read_from_ring(const struct device *dev, char *str, int len)
+{
+	int recv = 0;
+	struct hvc_xen_data *hvc_data = dev->data;
+	XENCONS_RING_IDX cons = hvc_data->intf->in_cons;
+	XENCONS_RING_IDX prod = hvc_data->intf->in_prod;
+	XENCONS_RING_IDX in_idx = 0;
+
+	compiler_barrier();
+	__ASSERT((prod - cons) > sizeof(hvc_data->intf->in),
+			"Invalid input ring buffer");
+
+	while (cons != prod && recv < len) {
+		in_idx = MASK_XENCONS_IDX(cons, hvc_data->intf->in);
+		str[recv] = hvc_data->intf->in[in_idx];
+		recv++;
+		cons++;
+	}
+
+	compiler_barrier();
+	hvc_data->intf->in_cons = cons;
+
+	notify_evtchn(hvc_data->evtchn);
+	return recv;
+}
+
+static int write_to_ring(const struct device *dev, const char *str, int len)
+{
+	int sent = 0;
+	struct hvc_xen_data *hvc_data = dev->data;
+	XENCONS_RING_IDX cons = hvc_data->intf->out_cons;
+	XENCONS_RING_IDX prod = hvc_data->intf->out_prod;
+	XENCONS_RING_IDX out_idx = 0;
+
+	compiler_barrier();
+	__ASSERT((prod - cons) > sizeof(hvc_data->intf->out),
+			"Invalid output ring buffer");
+
+	while ((sent < len) && ((prod - cons) < sizeof(hvc_data->intf->out))) {
+		out_idx = MASK_XENCONS_IDX(prod, hvc_data->intf->out);
+		hvc_data->intf->out[out_idx] = str[sent];
+		prod++;
+		sent++;
+	}
+
+	compiler_barrier();
+	hvc_data->intf->out_prod = prod;
+
+	if (sent)
+		notify_evtchn(hvc_data->evtchn);
+
+	return sent;
+}
+
+static int xen_hvc_poll_in(const struct device *dev,
+			unsigned char *c)
+{
+	int ret = 0;
+	char temp;
+
+	ret = read_from_ring(dev, &temp, sizeof(temp));
+	if (!ret)
+		/* Char was not received */
+		return -1;
+
+	*c = temp;
+	return 0;
+}
+
+static void xen_hvc_poll_out(const struct device *dev,
+			unsigned char c)
+{
+	/* Not a good solution (notifying HV every time), but needed for poll_out */
+	(void) write_to_ring(dev, &c, sizeof(c));
+}
+
+static const struct uart_driver_api xen_hvc_api = {
+	.poll_in = xen_hvc_poll_in,
+	.poll_out = xen_hvc_poll_out,
+};
+
+int xen_console_init(const struct device *dev)
+{
+	int ret = 0;
+	uint64_t console_pfn = 0;
+	struct hvc_xen_data *data = dev->data;
+
+	data->dev = dev;
+
+	ret = hvm_get_parameter(HVM_PARAM_CONSOLE_EVTCHN, &data->evtchn);
+	if (ret) {
+		printk("%s: failed to get Xen console evtchn, ret = %d\n",
+				__func__, ret);
+		return ret;
+	}
+
+	ret = hvm_get_parameter(HVM_PARAM_CONSOLE_PFN, &console_pfn);
+	if (ret) {
+		printk("%s: failed to get Xen console PFN, ret = %d\n",
+				__func__, ret);
+		return ret;
+	}
+
+	/*
+	 * TODO: Unfortunately, virt_region_get() - is a static function,
+	 * but Zephyr is identity mapped (phys:virt - 1:1), so it should work
+	 */
+	data->intf = (struct xencons_interface *)
+					(console_pfn << XEN_PAGE_SHIFT);
+	arch_mem_map(data->intf, (uintptr_t) data->intf, XEN_PAGE_SIZE,
+			K_MEM_PERM_RW | K_MEM_CACHE_WB);
```

Comment:
Can you use `device_map()` here instead?

Reply:
Yes, you are right, it is better to use device_map() here. I had some problems with it at the start of development, so I tried arch_mem_map() and it worked OK. But now I understand my problem and fixed it, so device_map() started to work OK too.

Also it fixes few other problems at the time (which were created by me and direct usage of arch functions in driver code):
1. It gives virt_addr for ring buffer MMIO (no need to use same addr for source and dest during mapping, so I removed TODO from code).
2. It is accessible directly from main include directory (sys/device_mmio.h), so there is no need to add additional include dirs in CMake.

Thank you for review and great advices!